flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/DistributedRuntimeUDFContext.java

```diff
@@ -41,10 +44,14 @@
 public class DistributedRuntimeUDFContext extends AbstractRuntimeUDFContext {
 
 	private final HashMap<String, BroadcastVariableMaterialization<?, ?>> broadcastVars = new HashMap<String, BroadcastVariableMaterialization<?, ?>>();
-	
+
+	private final Map<String, Set<? extends ExternalResourceInfo>> externalResources;
+
 	public DistributedRuntimeUDFContext(TaskInfo taskInfo, ClassLoader userCodeClassLoader, ExecutionConfig executionConfig,
-											Map<String, Future<Path>> cpTasks, Map<String, Accumulator<?,?>> accumulators, MetricGroup metrics) {
+										Map<String, Future<Path>> cpTasks, Map<String, Accumulator<?, ?>> accumulators,
+										MetricGroup metrics, Map<String, Set<? extends ExternalResourceInfo>> externalResources) {
 		super(taskInfo, userCodeClassLoader, executionConfig, accumulators, cpTasks, metrics);
+		this.externalResources = externalResources;
```

Comment:
If you look at the callers of this constructor, it seems that most of the arguments are derived from an `Environment`. Maybe we can just pass in an `Environment` to reduce the amount of arguments.

Alternatively, if we do not want to expose the entire `Environment` to this class, we can consider a builder/factory that takes an `Environment` and generates the constructor arguments.

Reply:
I like the idea to have a builder/factory that takes an `Environment`. However, it seems the `RuntimeContext` should not couple with `Environment`. `RuntimeContext` is in Function level while `Environment` is TM level. Flink does not ensure that the TM always exists when trigger a Function.